infra/src/main/kotlin/fi/fta/geoviite/infra/tracklayout/LocationTrackSpatialCache.kt

```diff
@@ -0,0 +1,128 @@
+
+import com.github.davidmoten.rtree2.RTree
+import com.github.davidmoten.rtree2.geometry.Geometries
+import com.github.davidmoten.rtree2.geometry.Rectangle
+import fi.fta.geoviite.infra.common.IntId
+import fi.fta.geoviite.infra.common.LayoutContext
+import fi.fta.geoviite.infra.common.RowVersion
+import fi.fta.geoviite.infra.math.IPoint
+import fi.fta.geoviite.infra.math.lineLength
+import java.util.concurrent.ConcurrentHashMap
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.stereotype.Service
+
+@Service
+class LocationTrackSpatialCache
+@Autowired
+constructor(
+    private val locationTrackService: LocationTrackService,
+    private val locationTrackDao: LocationTrackDao,
+    private val alignmentDao: LayoutAlignmentDao,
+) {
+    private val caches: ConcurrentHashMap<LayoutContext, ContextCache> = ConcurrentHashMap()
+
+    fun get(context: LayoutContext): ContextCache {
+        val all = locationTrackService.list(context).associateBy { it.id as IntId }
+        return caches.compute(context) { _, cache -> refresh(cache ?: newCache(), all) }
+            ?: error("Cache should have been created")
+    }
+
+    private fun newCache(): ContextCache = ContextCache(locationTrackDao::fetch, alignmentDao::fetch)
+
+    private fun refresh(cache: ContextCache, newTracks: Map<IntId<LocationTrack>, LocationTrack>): ContextCache {
+        val (staleTracks, currentTracks) =
+            cache.items
+                .asSequence()
+                .partition { (id, track) -> newTracks[id]?.alignmentVersion != track.alignmentVersion }
+                .let { (stale, current) ->
+                    stale.associate { it.key to it.value } to current.associate { it.key to it.value }
+                }
+
+        // Remove all stale items (tracks that have been removed or updated)
+        var newNet =
+            staleTracks
+                .flatMap { (_, entry) -> entry.segmentData }
+                .fold(cache.network) { net, (segment, rect) -> net.delete(segment, rect) }
+
+        // Add all new items (tracks that have been added or updated)
+        fun addEntry(track: LocationTrack) =
+            createEntry(track, alignmentDao.fetch(track.getAlignmentVersionOrThrow())).also { entry ->
+                newNet = entry.segmentData.fold(newNet) { net, (segment, rect) -> net.add(segment, rect) }
+            }
+
+        val newItems = newTracks.map { (id, track) -> id to (currentTracks[id] ?: addEntry(track)) }
+
+        return ContextCache(locationTrackDao::fetch, alignmentDao::fetch, newNet, newItems.toMap())
+    }
+}
+
+data class SpatialCacheSegment(
+    val locationTrackVersion: LayoutRowVersion<LocationTrack>,
+    val alignmentVersion: RowVersion<LayoutAlignment>,
+    val segment: LayoutSegment,
+)
+
+data class SpatialCacheEntry(
+    val alignmentVersion: RowVersion<LayoutAlignment>,
+    val segmentData: List<Pair<SpatialCacheSegment, Rectangle>>,
+)
+
+data class CacheHit(
```

Comment:
Olio joita cache antaa ulos

Reply:
Koska tämä spatiaalikakku on tällä hetkellä sijaintiraidespesifinen, niin lisäisin itse tähän nimeen `LocationTrack`-termin (tyyliin `LocationTrackCacheHit`.) Huomasin ylempiä tiedostoja lukiessani että jäin miettimään minkäköhänlaisia osumia `CacheHit`-tyyppinen muuttuja käytännössä sisältää (tai siis voisiko se sisältää muutakin kuin sijaintiraiteita.)